File: samples/text_detection/dbnet.cpp

/*
 * Copyright (C) 2025 Vastai-tech Company.
 * All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
#include "common/cmdline.hpp"
#include "common/file_system.hpp"
#include "common/utils.hpp"
#include "dbnet_detector/dbnet_detector.hpp"
#include "opencv2/opencv.hpp"

cmdline::parser ArgumentParser(int argc, char **argv) {
  cmdline::parser args;
  args.add<std::string>(
      "model_prefix", 'm', "model prefix of the model suite files", false,
      "/opt/vastai/vaststreamx/data/models/"
      "dbnet_resnet50_vd-int8-kl_divergence-1_3_736_1280-vacc/mod");
  args.add<std::string>("hw_config", '\0', "hw-config file of the model suite",
                        false);
  args.add<std::string>("vdsp_params", '\0', "vdsp preprocess parameter file",
                        false, "../data/configs/dbnet_rgbplanar.json");
  args.add<uint32_t>("device_id", 'd', "device id to run", false, 0);
  args.add<float>("threshold", '\0', "threshold for detection", false, 0.3);
  args.add<float>("box_threshold", '\0', "threshold for boxes", false, 0.6);
  args.add<float>("box_unclip_ratio", '\0', "unclip ratio", false, 1.5);
  args.add<bool>("use_polygon_score", '\0', "use_polygon_score in postprocess",
                 false, false);
  args.add<std::string>(
      "elf_file", '\0', "elf file path", false,
      "/opt/vastai/vaststreamx/data/elf/find_contours_ext_op");
  args.add<std::string>("input_file", '\0', "input image file", false,
                        "../data/images/detect.jpg");
  args.add<std::string>("output_file", '\0', "output image file", false, "");
  args.add<std::string>("dataset_filelist", '\0', "input dataset filelist",
                        false, "");
  args.add<std::string>("dataset_root", '\0', "input dataset root", false, "");
  args.add<std::string>("dataset_output_folder", '\0', "dataset output folder",
                        false, "");
  args.parse_check(argc, argv);
  return args;
}

int main(int argc, char **argv) {
  auto args = ArgumentParser(argc, argv);
  const int batch_size = 1;

  vsx::DBnetDetector detector(args.get<std::string>("model_prefix"),
                              args.get<std::string>("vdsp_params"),
                              args.get<std::string>("elf_file"), batch_size,
                              args.get<uint32_t>("device_id"));
  detector.SetThreshold(args.get<float>("threshold"));
  detector.SetBoxThreshold(args.get<float>("box_threshold"));
  detector.SetPolygonScoreUsage(args.get<bool>("use_polygon_score"));
  detector.SetBoxUnclipRatio(args.get<float>("box_unclip_ratio"));
  auto image_format = detector.GetFusionOpIimageFormat();

  if (args.get<std::string>("dataset_filelist").empty()) {
    vsx::Image vsx_image;
    vsx::MakeVsxImage(args.get<std::string>("input_file"), vsx_image,
                      image_format);
    auto res = detector.Process(vsx_image);
    if (res.GetSize() == 0) {
      std::cout << "No object detected in image.\n";
    } else {
      // print result and draw rectangle
      auto cv_image = cv::imread(args.get<std::string>("input_file"));
      int obj_count = res.Shape()[0];
      const float *data = res.Data<float>();
      for (int i = 0; i < obj_count; i++) {
        std::cout << "index:" << i << ", score:" << data[i * 9 + 0]
                  << ",bbox:[ [" << static_cast<int>(data[i * 9 + 1]) << " "
                  << static_cast<int>(data[i * 9 + 2]) << "] ["
                  << static_cast<int>(data[i * 9 + 3]) << " "
                  << static_cast<int>(data[i * 9 + 4]) << "] ["
                  << static_cast<int>(data[i * 9 + 5]) << " "
                  << static_cast<int>(data[i * 9 + 6]) << "] ["
                  << static_cast<int>(data[i * 9 + 7]) << " "
                  << static_cast<int>(data[i * 9 + 8]) << "] ]\n";
        cv::line(cv_image, cv::Point2f(data[i * 9 + 1], data[i * 9 + 2]),
                 cv::Point2f(data[i * 9 + 3], data[i * 9 + 4]),
                 cv::Scalar(0, 0, 255));
        cv::line(cv_image, cv::Point2f(data[i * 9 + 3], data[i * 9 + 4]),
                 cv::Point2f(data[i * 9 + 5], data[i * 9 + 6]),
                 cv::Scalar(0, 0, 255));
        cv::line(cv_image, cv::Point2f(data[i * 9 + 5], data[i * 9 + 6]),
                 cv::Point2f(data[i * 9 + 7], data[i * 9 + 8]),
                 cv::Scalar(0, 0, 255));
        cv::line(cv_image, cv::Point2f(data[i * 9 + 1], data[i * 9 + 2]),
                 cv::Point2f(data[i * 9 + 7], data[i * 9 + 8]),
                 cv::Scalar(0, 0, 255));
      }

      // save file
      if (!args.get<std::string>("output_file").empty()) {
        cv::imwrite(args.get<std::string>("output_file"), cv_image);
      }
    }

  } else {
    std::vector<std::string> filelist =
        vsx::ReadFileList(args.get<std::string>("dataset_filelist"));
    auto dataset_root = args.get<std::string>("dataset_root");
    for (auto &file : filelist) {
      auto fullname = file;
      if (!dataset_root.empty()) fullname = dataset_root + "/" + file;
      std::cout << fullname << std::endl;
      vsx::Image vsx_image;
      vsx::MakeVsxImage(fullname, vsx_image, image_format);
      auto res = detector.Process(vsx_image);
      vsx::Tensor output_tensor;
      if (res.GetSize() == 0) {
        output_tensor = vsx::Tensor(vsx::TShape({0}), vsx::Context::CPU(),
                                    vsx::TypeFlag::kUint32);
      } else {
        int obj_count = res.Shape()[0];
        output_tensor =
            vsx::Tensor(vsx::TShape({obj_count, 4, 2}), vsx::Context::CPU(),
                        vsx::TypeFlag::kUint32);
        const float *src = res.Data<float>();
        uint32_t *dst = output_tensor.MutableData<uint32_t>();
        for (int i = 0; i < obj_count; i++) {
          for (int j = 0; j < 8; j++) {
            dst[i * 8 + j] = static_cast<uint32_t>(src[i * 9 + j + 1]);
          }
        }
      }

      std::filesystem::path p(fullname);
      auto output_file = args.get<std::string>("dataset_output_folder") + "/" +
                         p.stem().string() + ".npz";
      std::cout << output_file << std::endl;
      std::unordered_map<std::string, vsx::Tensor> output_map;
      output_map["output_0"] = output_tensor;
      vsx::SaveTensorMap(output_file, output_map);
    }
  }

  return 0;
}

File: samples/text_detection/dbnet.py
import os
import sys

current_file_path = os.path.dirname(os.path.abspath(__file__))
common_path = os.path.join(current_file_path, "../..")
sys.path.append(common_path)

import numpy as np
import cv2
import argparse
from dbnet_detector.dbnet_detector import DbnetDetector
import vaststreamx as vsx
import common.utils as utils


def argument_parser():
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "-m",
        "--model_prefix",
        default="/opt/vastai/vaststreamx/data/models/dbnet_resnet50_vd-int8-kl_divergence-1_3_736_1280-vacc/mod",
        help="model prefix of the model suite files",
    )
    parser.add_argument(
        "--hw_config",
        default="",
        help="hw-config file of the model suite",
    )
    parser.add_argument(
        "--vdsp_params",
        default="./data/configs/dbnet_rgbplanar.json",
        help="vdsp preprocess parameter file",
    )
    parser.add_argument(
        "-d",
        "--device_id",
        default=0,
        type=int,
        help="device id to run",
    )
    parser.add_argument(
        "--elf_file",
        default="/opt/vastai/vaststreamx/data/elf/find_contours_ext_op",
        help="input file",
    )
    parser.add_argument(
        "--input_file",
        default="../../../data/images/detect.jpg",
        help="input file",
    )
    parser.add_argument(
        "--output_file",
        default="./dbnet_result.jpg",
        help="output file",
    )
    parser.add_argument(
        "--dataset_filelist",
        default="",
        help="input dataset image list",
    )
    parser.add_argument(
        "--dataset_root",
        default="",
        help="input dataset root",
    )
    parser.add_argument(
        "--dataset_output_folder",
        default="",
        help="dataset output folder",
    )
    args = parser.parse_args()
    return args


if __name__ == "__main__":
    args = argument_parser()
    batch_size = 1
    detector = DbnetDetector(
        args.model_prefix,
        args.vdsp_params,
        batch_size,
        args.device_id,
        args.hw_config,
        elf_file=args.elf_file,
    )
    input_shape = detector.input_shape[0]
    image_format = detector.get_fusion_op_iimage_format()

    if args.dataset_filelist == "":
        cv_image = cv2.imread(args.input_file)
        assert cv_image is not None, f"Read image failed:{args.input_file}"
        vsx_image = utils.cv_bgr888_to_vsximage(cv_image, image_format, args.device_id)
        [bboxes, scores] = detector.process(vsx_image)
        for i in range(len(bboxes)):
            print(
                f"index:{i}, score:{scores[i]},bbox:[{bboxes[i][0]},{bboxes[i][1]},{bboxes[i][2]},{bboxes[i][3]}]"
            )
        if args.output_file != "":
            for bbox in bboxes:
                for i in range(len(bbox)):
                    t = (i + 1) % len(bbox)
                    pt1 = (bbox[i][0], bbox[i][1])
                    pt2 = (bbox[t][0], bbox[t][1])
                    cv2.line(cv_image, pt1, pt2, color=(0, 0, 255))
            cv2.imwrite(args.output_file, cv_image)
    else:
        filelist = []
        with open(args.dataset_filelist, "rt") as f:
            filelist = f.readlines()
        for filename in filelist:
            filename = os.path.join(args.dataset_root, filename.replace("\n", ""))
            print(filename)
            cv_image = cv2.imread(filename)
            assert cv_image is not None, f"Read image failed:{filename}"
            vsx_image = utils.cv_bgr888_to_vsximage(
                cv_image, image_format, args.device_id
            )
            [bboxes, scores] = detector.process(vsx_image)
            basename, _ = os.path.splitext(os.path.basename(filename))
            npz_file = os.path.join(args.dataset_output_folder, basename + ".npz")
            np.savez(npz_file, output_0=bboxes)
